File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int na = nums1.length;
        int nb = nums2.length;

        int n = na + nb;
        if (n % 2 == 1) {
            return helper(n / 2, nums1, 0, na - 1, nums2, 0, nb - 1);
        } else {
            return (double)(helper(n / 2, nums1, 0, na - 1, nums2, 0, nb - 1) + helper(n / 2 - 1, nums1, 0, na - 1, nums2, 0, nb - 1)) / 2;
        }
    }

    private int helper(int k, int[] nums1, int aLeft, int aRight, int[] nums2, int bLeft, int bRight) {
        if (aRight < aLeft) {
            return nums2[k - aLeft];
        }
        if (bRight < bLeft) {
            return nums1[k - bLeft];
        }

        int aMid = (aLeft + aRight) / 2;
        int bMid = (bLeft + bRight) / 2;
        int aValue = nums1[aMid];
        int bValue = nums2[bMid];

        if (aMid + bMid < k) {
            if (aValue > bValue) {
                return helper(k, nums1, aLeft, aRight, nums2, bMid + 1, bRight);
            } else {
                return helper(k, nums1, aMid + 1, aRight, nums2, bLeft, bRight);
            }
        } else {
            if (aValue > bValue) {
                return helper(k, nums1, aLeft, aMid - 1, nums2, bLeft, bRight);
            } else {
                return helper(k, nums1, aLeft, aRight, nums2, bLeft, bMid - 1);
            }            
        }
    }
}

File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int m = nums1.size();
        int n = nums2.size();

        if ((m + n) % 2 == 1) {
            return findKthSmallest(nums1, 0, m - 1, nums2, 0, n - 1, (m + n) / 2 + 1);
        } else {
            int first = findKthSmallest(nums1, 0, m - 1, nums2, 0, n - 1, (m + n) / 2);
            int second = findKthSmallest(nums1, 0, m - 1, nums2, 0, n - 1, (m + n) / 2 + 1);

            return (double)(first + second) / 2;
        }
    }

private:
    int findKthSmallest(vector<int>& A, int Astart, int Aend, 
                        vector<int>& B, int Bstart, int Bend, int k) {
        if (Astart > Aend) {
            return B[Bstart + k - 1];
        }
        if (Bstart > Bend) {
            return A[Astart + k - 1];
        }
        if (k == 1) {
            return min(A[Astart], B[Bstart]);
        }

        int i = min((int)A.size() - Astart, k / 2);
        int j = min((int)B.size() - Bstart, k / 2);

        if (A[Astart + i - 1] > B[Bstart + j - 1]) {
            return findKthSmallest(A, Astart, Aend, B, Bstart + j, Bend, k - j);
        } else {
            return findKthSmallest(A, Astart + i, Aend, B, Bstart, Bend, k - i);
        }
    } 
};